step-core/src/main/java/step/parameter/ParameterManager.java

```diff
@@ -34,37 +31,79 @@
 import step.core.encryption.EncryptionManagerException;
 import step.core.objectenricher.ObjectPredicate;
+import step.core.plugins.exceptions.PluginCriticalException;
 
 public class ParameterManager {
 	
 	public static final String RESET_VALUE = "####change me####";
+	public static final String PROTECTED_VALUE = "******";
 
 	private static Logger logger = LoggerFactory.getLogger(ParameterManager.class);
 	
 	private Accessor<Parameter> parameterAccessor;
 	private EncryptionManager encryptionManager;
 
 	private String defaultScriptEngine;
+	private final DynamicBeanResolver dynamicBeanResolver;
 
-	public ParameterManager(Accessor<Parameter> parameterAccessor, EncryptionManager encryptionManager, Configuration configuration) {
-		this(parameterAccessor, encryptionManager, configuration.getProperty("tec.activator.scriptEngine", Activator.DEFAULT_SCRIPT_ENGINE));
+	public ParameterManager(Accessor<Parameter> parameterAccessor, EncryptionManager encryptionManager, Configuration configuration, DynamicBeanResolver dynamicBeanResolver) {
+		this(parameterAccessor, encryptionManager, configuration.getProperty("tec.activator.scriptEngine", Activator.DEFAULT_SCRIPT_ENGINE), dynamicBeanResolver);
 	}
 
-	public ParameterManager(Accessor<Parameter> parameterAccessor, EncryptionManager encryptionManager, String defaultScriptEngine) {
+	public ParameterManager(Accessor<Parameter> parameterAccessor, EncryptionManager encryptionManager, String defaultScriptEngine, DynamicBeanResolver dynamicBeanResolver) {
 		this.parameterAccessor = parameterAccessor;
 		this.encryptionManager = encryptionManager;
 		this.defaultScriptEngine = defaultScriptEngine;
+		this.dynamicBeanResolver = dynamicBeanResolver;
 	}
 
 	public static ParameterManager copy(ParameterManager from, Accessor<Parameter> parameterAccessor){
-		return new ParameterManager(parameterAccessor, from.encryptionManager, from.defaultScriptEngine);
+		return new ParameterManager(parameterAccessor, from.encryptionManager, from.defaultScriptEngine, from.dynamicBeanResolver);
+	}
+
+	public Parameter save(Parameter newParameter, Parameter sourceParameter, String modificationUser) {
+		if (isProtected(newParameter) && newParameter.getValue() != null && newParameter.getValue().isDynamic()) {
+			throw new ParameterManagerException("Protected parameters do not support values with dynamic expression.");
+		}
+
+		ParameterScope scope = newParameter.getScope();
+		if(scope != null && scope.equals(ParameterScope.GLOBAL) && newParameter.getScopeEntity() != null) {
+			throw new ParameterManagerException("Scope entity cannot be set for parameters with GLOBAL scope.");
+		}
+
+		if(sourceParameter != null) {
+			// the parameter has been updated but the value hasn't been changed
+			DynamicValue<String> newParameterValue = newParameter.getValue();
+			if(newParameterValue != null && !newParameterValue.isDynamic() && newParameterValue.get().equals(PROTECTED_VALUE)) {

```

Comment:
Not sure if this is still the optimal implementation. It forces the clients to set the value to "******" in order for the value not to change when updating a parameter.

Reply:
@jeromecomte so you would simply remove this if block and move the line of code 
```
newParameter.setValue(sourceParameter.getValue());
```
to the following if bloc 
```
if(isProtected(sourceParameter)) {
    // protected value should not be changed
    newParameter.setProtectedValue(true);
    newParameter.setValue(sourceParameter.getValue());
}
```